services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala

```diff
@@ -178,14 +178,73 @@ trait MetadataDatabaseAccess {
       queryParameters.startDate.map(_.toSystemTimestamp), queryParameters.endDate.map(_.toSystemTimestamp))
 
-    workflowSummaryCount flatMap { count =>
-      workflowSummaries map { workflows =>
-        (WorkflowQueryResponse(workflows.toSeq map { workflow =>
-          MetadataService.WorkflowQueryResult(id = workflow.workflowExecutionUuid, name = workflow.workflowName, status = workflow.workflowStatus, start = workflow.startTimestamp map { _.toSystemOffsetDateTime }, end = workflow.endTimestamp map { _.toSystemOffsetDateTime })
-        }),
-          //only return metadata if page is defined
-          queryParameters.page map { _ => QueryMetadata(queryParameters.page, queryParameters.pageSize, Option(count)) })
+    def queryMetadata(count: Int): Option[QueryMetadata] = {
+      queryParameters.page map { _ => QueryMetadata(queryParameters.page, queryParameters.pageSize, Option(count)) }
+    }
+
+    def summariesToQueryResults(workflows: Traversable[WorkflowMetadataSummaryEntry]): Future[Traversable[MetadataService.WorkflowQueryResult]] = {
+      val mapped: Traversable[Future[MetadataService.WorkflowQueryResult]] = workflows map { summaryToQueryResult }
+      val result: Future[Traversable[MetadataService.WorkflowQueryResult]] = Future.sequence(mapped)
+      result
+    }
+
+    def summaryToQueryResult(workflow: WorkflowMetadataSummaryEntry): Future[MetadataService.WorkflowQueryResult] = {
+
+      def queryResult(labels: Map[String, String], parentId: Option[String]): MetadataService.WorkflowQueryResult = {
+        MetadataService.WorkflowQueryResult(
+          id = workflow.workflowExecutionUuid,
+          name = workflow.workflowName,
+          status = workflow.workflowStatus,
+          start = workflow.startTimestamp map { _.toSystemOffsetDateTime },
+          end = workflow.endTimestamp map { _.toSystemOffsetDateTime },
+          labels = if (labels.nonEmpty) { Some(labels) } else None,
+          parentWorkflowId = parentId
+        )
+      }
+
+      def metadataEntriesToValue(entries: Seq[MetadataEntry]): Option[String] = {
+        entries.headOption.flatMap(_.metadataValue.toRawStringOption)
+      }
+
+      def keyToMetadataValue(key: String): Future[Option[String]] = {
+        val metadataEntries: Future[Seq[MetadataEntry]] = metadataDatabaseInterface.queryMetadataEntries(workflow.workflowExecutionUuid, key)
+        metadataEntries map { entries =>
+          metadataEntriesToValue(entries)
+        }
+      }
+
+      def getWorkflowLabels: Future[Map[String, String]] = {
+        if (queryParameters.additionalKeys.contains(WorkflowMetadataKeys.Labels)) {
+          metadataDatabaseInterface.getWorkflowLabels(workflow.workflowExecutionUuid)
+        } else {
+          Future.successful(Map.empty)
+        }
+      }
+
+      val getParentWorkflowId: Future[Option[String]] = {
```

Comment:
As mentioned above, I think it's gonna be much nicer and more scalable as we add more possible "additional fields" to make these part of the original query rather than added on as subsequent queries to the DB (since metadata is a *big* table)